funsor/terms.py

```diff
@@ -376,19 +375,37 @@ def reduce(self, op, reduced_vars=None):
         Reduce along all or a subset of inputs.
 
         :param op: A reduction operation.
-        :type op: ~funsor.ops.AssociativeOp
+        :type op: ~funsor.ops.AssociativeOp or ~funsor.ops.ReductionOp
         :param reduced_vars: An optional input name or set of names to reduce.
             If unspecified, all inputs will be reduced.
         :type reduced_vars: str, Variable, or set or frozenset thereof.
         """
-        assert isinstance(op, AssociativeOp)
+        assert isinstance(op, (AssociativeOp, ops.ReductionOp))
+
         # Eagerly convert reduced_vars to appropriate things.
         if reduced_vars is None:
             # Empty reduced_vars means "reduce over everything".
             reduced_vars = frozenset(Variable(k, v) for k, v in self.inputs.items())
         else:
             reduced_vars = _convert_reduced_vars(reduced_vars, self.inputs)
         assert isinstance(reduced_vars, frozenset), reduced_vars
+
+        # Attempt to convert ReductionOp to AssociativeOp.
+        if isinstance(op, ops.ReductionOp):
```

Comment:
Do we want this standardization logic to run under all interpretations?

Reply:
yes, I see `.reduce()` as lightweight syntax around actual syntax including `Reduce`. Note it is nonsensical to create a lazy `Reduce` of a `ReductionOp`, the way we have defined `Reduce`.